controller/sharding/sharding.go

```diff
@@ -23,31 +33,83 @@
 	}
 }
 
-func GetClusterFilter(replicas int, shard int) func(c *v1alpha1.Cluster) bool {
-	return func(c *v1alpha1.Cluster) bool {
-		clusterShard := 0
-		//  cluster might be nil if app is using invalid cluster URL, assume shard 0 in this case.
+func GetDistributionFunction(kubernetesClient *kubernetes.Clientset, settingsMgr *settings.SettingsManager) func(c *v1alpha1.Cluster) int {
+	filterFunctionName := env.StringFromEnv(common.EnvControllerShardingAlgorithm, "legacy")
+	distributionFunction := GetShardByIdUsingHashDistributionFunction()
+	log.Infof("Using filter function:  %s", filterFunctionName)
+	switch {
+	case filterFunctionName == "hash":
+		distributionFunction = GetShardByIndexModuloReplicasCountDistributionFunction(settingsMgr, kubernetesClient)
+	case filterFunctionName == "legacy":
+		distributionFunction = GetShardByIdUsingHashDistributionFunction()
+	}
+	return distributionFunction
+}
+
+func GetShardByIndexModuloReplicasCountDistributionFunction(settingsMgr *settings.SettingsManager, kubeClientset *kubernetes.Clientset) DistributionFunction {
+	namespace := settingsMgr.GetNamespace()
+	db := db.NewDB(namespace, settingsMgr, kubeClientset)
+	replicas := env.ParseNumFromEnv(common.EnvControllerReplicas, 0, 0, math.MaxInt32)
+	ctx := context.Background()
+	clustersList, dbErr := db.ListClusters(ctx)
+	if dbErr != nil {
+		log.Warnf("Error while querying clusters list from database: %v", dbErr)
+	}
+	clusters := clustersList.Items
+	sort.Slice(clusters, func(i, j int) bool {
+		return clusters[i].ID < clusters[j].ID
+	})
+	log.Debugf("ClustersList has %d items", len(clusters))
+	clusterById := make(map[string]v1alpha1.Cluster)
+	clusterIndexdByClusterId := make(map[string]int)
+	for i, cluster := range clusters {
+		log.Debugf("Adding cluster with id=%s and name=%s to clusterIndexdByClusterId map", cluster.ID, cluster.Name)
+		clusterById[cluster.ID] = cluster
+		clusterIndexdByClusterId[cluster.ID] = i
+	}
+
+	return func(c *v1alpha1.Cluster) int {
 		if c != nil {
-			if c.Shard != nil {
-				clusterShard = int(*c.Shard)
-			} else {
-				clusterShard = GetShardByID(c.ID, replicas)
+			clusterIndex, ok := clusterIndexdByClusterId[c.ID]
+			if !ok {
+				log.Infof("Cluster with id=%s not found in clusterIndexdByClusterId map", c.ID)
 			}
+			shard := int(clusterIndex % replicas)
+			log.Infof("Cluster with id=%s will be processed by shard %d", c.ID, shard)
+			return shard
+		}
+		return 0
+	}
+}
+
+func GetShardByIdUsingHashDistributionFunction() DistributionFunction {
+	replicas := env.ParseNumFromEnv(common.EnvControllerReplicas, 0, 0, math.MaxInt32)
+	return func(c *v1alpha1.Cluster) int {
+		if c == nil {
+			log.Infof("Calculating cluster shard for cluster id: %v", c)
+			return 0
+		}
+		id := c.ID
+		log.Infof("Calculating cluster shard for cluster id: %s", id)
+		if id == "" {
+			return 0
+		} else {
+			h := fnv.New32a()
+			_, _ = h.Write([]byte(id))
+			shard := int32(h.Sum32() % uint32(replicas))
```

Comment:
This needs a change in `env.go` avoid using  `strconv.Atoi` :

```
func ParseNumFromEnv(env string, defaultValue, min, max int) int {
	str := os.Getenv(env)
	if str == "" {
		return defaultValue
	}
	num, err := strconv.ParseInt(str, 10, 32)
	if err != nil {
		log.Warnf("Could not parse '%s' as a number from environment %s", str, env)
		return defaultValue
	}
	if int(num) < min {
		log.Warnf("Value in %s is %d, which is less than minimum %d allowed", env, num, min)
		return defaultValue
	}
	if int(num) > max {
		log.Warnf("Value in %s is %d, which is greater than maximum %d allowed", env, num, max)
		return defaultValue
	}
	return int(num)
}

```

By using something like this, it fixes the issue, but unit tests are failing then because of max values of `int` versus max value of `int64`

Reply:
Issue https://github.com/argoproj/argo-cd/issues/13136 has been created for this.